### PR TITLE
core1.0: initial spec for the core1.{sub,pub,set} messages

### PR DESCRIPTION
**Pull request by majewsky**
This is... completely HUGE... and kind of immortal. I definitely need to reread this tomorrow or so when my mind has recovered from this writing session.

@hkgit03 Feel free to take a look. However, this is so big that we'd probably be better served by reviewing this in person. After writing this down, I have some reservations regarding all the complexity that we're introducing here.

In related news, I've moved the todolist-in-a-comment from the bottom of `core1.0` into my notes file. I also added some notes that we should also go through in our next design session.

### CODE REVIEW

File: spec/core/1.0.md [hunk 1]
```diff
@@ -417,15 +417,124 @@ If the client finds that the server has not agreed to a module or capability tha
...
+- not the name of a property,
+- the name of a property in a module to whose usage the server has not agreed,
+- the name of a property which is restricted to a capability to which the server has not agreed, or
+- the name of a property to which the recipient of the `core1.pub` message has not subscribed.
+
+Moreover, a `core1.pub` message is invalid if any of its even-numbered arguments is a property value which is not valid for the property in question according to the specification defining the property.
+
+A `core1.pub` message can be sent by the server at any time to notify the client of changes to the values of properties that the client has subscribed to.
+Moreover, a `core1.pub` message MUST be sent by the server immediately upon receiving a valid `core1.sub` or `core1.set` message.
+This message MUST contain the values of all properties named in the `core1.pub` or `core1.set` message, in the same order.
```
Comment (laerling) [inline spec/core/1.0.md]: I think you've got a typo: Expected `core1.sub`, found `core1.pub`.

File: spec/core/1.0.md [hunk 2]
```diff
@@ -417,15 +417,124 @@ If the client finds that the server has not agreed to a module or capability tha
...
+For `core1.pub` messages, this applies regardless of whether the client sending the message subscribes to properties for the first time, or whether subscriptions to these properties already existed for this client.
```
Comment (laerling) [inline spec/core/1.0.md]: Same typo as above, if any.

File: spec/core/1.0.md [hunk 3]
```diff
@@ -417,15 +417,124 @@ If the client finds that the server has not agreed to a module or capability tha
...
+--> invalid; missing a property mentioned in the `core1.sub` message
+
+(core1.sub foo1.bar foo1.baz)
+(core1.pub foo1.bar hello-world foo1.baz "20" foo1.extra "10")
+--> invalid; references a property not mentioned in the `core1.sub` message
+
+(core1.sub foo1.bar foo1.baz)
+(core1.pub foo1.baz "20" foo1.bar hello-world)
+--> invalid; properties not listed in correct order
+```
```
Comment (laerling) [inline spec/core/1.0.md:497]: Missing edge case: The client asks for a property that doesn't exist.

Comment (laerling) [reply spec/core/1.0.md:497]: Nevermind. This falls under "property in a module to whose usage the server has not agreed" (line 453).
Maybe we should state explicitely though, that this includes properties which don't exist at all? (To make clear that every property that doesn't exist at all is a property in a potential module to which the server has not agreed)

File: spec/core/1.0.md [hunk 4]
```diff
@@ -417,15 +417,124 @@ If the client finds that the server has not agreed to a module or capability tha
...
+- the name of a property in a module to whose usage the server has not agreed, or
+- the name of a property which is restricted to a capability to which the server has not agreed.
+
+Notably, the message is not invalid if any of its even-numbered arguments is a value that is not permissible for the property in question.
+
+*Rationale:* For non-trivial properties, the client might not have enough information to decide whether a given value is valid.
+
+Setting a property implies subscribing to the property.
+That is, a `core1.set` message sent from a client to a server implies a `core1.sub` message from the same client to the same server that mentions the same properties that are mentioned in the `core1.set` message.
+This does not mean, however, that a separate `core1.pub` message must be sent to answer the implied `core1.pub` message.
```
Comment (laerling) [inline spec/core/1.0.md]: Expected "implied `core1.sub`", found "implied `core1.pub`".

**Review by laerling — CHANGES_REQUESTED** [review]
Mainly typos (crucial ones, though).
Additional note: Explicitely define that numbering arguments starts with 1, i. e. the first argument is argument 1, and so forth.

File: spec/core/1.0.md (hunk 3 above)
Comment (majewsky) [reply spec/core/1.0.md:497]: Yes, that's covered in lines 433 and 452.

**Review by laerling — APPROVED** [review]

**Pull request merged**